Add more comments to several files

dthaler opened the pull request:
Signed-off-by: Dave Thaler <dthaler@microsoft.com>

elazarg commented on src/assertions.cpp:
```diff
@@ -174,6 +175,7 @@ class AssertExtractor {
     }
 };
 
+/// Annotate the CFG by adding in assertions before every memory instruction.
```
This is incomplete. We add explicit assertions for all the preconditions of any instruction. For example, jump instructions are asserted not to compare numbers and pointers, or pointers to potentially distinct memory regions.

dthaler replied:
Updated

elazarg commented on src/main/check.cpp:
```diff
@@ -55,7 +55,7 @@ int main(int argc, char** argv) {
     app.add_option("--asm", asmfile, "Print disassembly to FILE")->type_name("FILE");
     std::string dotfile;
-    app.add_option("--dot", dotfile, "Export cfg to dot FILE")->type_name("FILE");
+    app.add_option("--dot", dotfile, "Export control-flow graph to dot FILE")->type_name("FILE");
```
I think that within this project, the term CFG should be obvious. I don't feel it helps reader to see three words everywhere for a term that is so fundamental to the project.

dthaler replied:
For such high level visible things as command line usage, having acronyms expanded makes this project more approachable to newcomers

elazarg commented on src/main/check.cpp:
```diff
@@ -107,16 +111,23 @@ int main(int argc, char** argv) {
     }
 
     if (domain == "zoneCrab") {
+        // Convert the instruction sequence to a control-flow graph.
```
... in a "passive", nondeterministic form.

dthaler replied:
Updated

elazarg commented on src/asm_cfg.cpp:
```diff
@@ -216,12 +224,20 @@ std::map<std::string, int> collect_stats(const cfg_t& cfg) {
     cfg_t cfg = to_nondet(det_cfg);
 
+    // Except when debugging, combine basic blocks where possible,
+    // i.e., where there is no jump to a later basic block.
```
I think it is more precise to say something like "combine chains of instructions to basic blocks" or something of the sort. The single instructions _can_ be thought of as trivial basic blocks, but the reader probably just sees them as instructions at this point.

Also, this is not exactly about jumps. A basic block is a sequence of instructions with a single entry point and a single exit point. Unconditional jumps do not necessarily break basic blocks, and fallthrough may or may not be concatenated - it depends on whether there exist a jump somewhere with the following instruction as its target.

dthaler replied:
Updated

elazarg commented on src/asm_cfg.cpp:
```diff
@@ -97,6 +100,9 @@ static vector<label_t> unique(const std::pair<T, T>& be) {
 
+/// Get a non-deterministic version of a control-flow graph,
+/// i.e., where instead of using if/else, both branches are taken
+/// simultaneously.
```
... and are replaced by `Assume` instructions immediately after the branch.

dthaler replied:
Updated.   (BTW, for explicit text suggestions github has nice tooling, click on the "+-" icon on the left end of the toolbar and insert your suggested modifications, which I can then accept via a github button on the resulting comment.)

elazarg commented on src/asm_cfg.cpp:
```diff
@@ -88,6 +90,7 @@ static Condition::Op reverse(Condition::Op op) {
 }
 
+/// Get the inverse of a given comparison condition.
 static Condition reverse(Condition cond) { return {.op = reverse(cond.op), .left = cond.left, .right = cond.right}; }
```
Perhaps we should rename these functions to `invert_comparison` if that is not clear.

dthaler replied:
That would be clearer.  Although this PR is so far only about comments.  If we want to do new renames in here too I can do that, let me know.  Otherwise could do renames in a subsequent PR.

elazarg replied:
Better not to mix. You are right.

elazarg commented on src/asm_syntax.hpp:
```diff
@@ -133,27 +140,32 @@ struct Deref {
     bool is_load{};
 };
 
+/// Legacy, no longer used.
```
It does exist in older eBPF programs, and prehaps even in newer programs under certain conditions. It is supported by the Linux runtime and is analyzed correctly (as a function call) in PREVAIL.

I'd say something like: A special instruction for checked access to packets; it is actually a function call, and analyzed as one, e.g. by scratching caller-saved registers after it is performed.

The relevant explanation in the linux verifier is [here](https://elixir.bootlin.com/linux/v5.9.8/source/kernel/bpf/verifier.c#L7315).

dthaler replied:
Updated

elazarg commented on src/asm_syntax.hpp:
```diff
@@ -133,27 +140,32 @@ struct Deref {
     int opcode{};
 };
 
+/// Special instruction added by analysis.
```
This is not an explanation.

> When a CFG is translated to its nondeterministic form, Conditional `Jump` instructions are replaced by two `Assume` instructions, immediately after the branch and before each jump target.

dthaler replied:
Updated

elazarg commented on src/asm_cfg.cpp:
```diff
@@ -138,6 +144,8 @@ static cfg_t to_nondet(const cfg_t& cfg) {
 
+/// Get the type of a given instruction.
+/// Most of these type names are also statistics header labels.
 static std::string instype(Instruction ins) {
```
It would be better to rename - `name_of_instruction_category()` or something like that.

dthaler replied:
Agree renames of lots of things would help.  But I was thinking to do only comments in this PR.  Let me know.

dthaler replied:
The one thing I did do in the update to this PR though is to add an enum for register numbers to reduce hard coding of magic numbers.  I think it makes it much more readable, let me know.

elazarg commented on src/asm_syntax.hpp:
```diff
@@ -164,10 +176,10 @@ enum { T_UNINIT = -6, T_MAP = -5, T_NUM = -4, T_CTX = -3, T_STACK = -2, T_PACKET
+    ctx,            ///< pointer to the context
+    packet,         ///< pointer to the packet
+    stack,          ///< pointer to the stack
+    shared,         ///< pointer to shared memory
```
These comments add very little IMO. If it is not clear that these are pointers, we should rename them to `ptr_ctx`, `ptr_packet` etc.

Also, "context" is vague  (for example the term might refer to pointer to a context in the verifier code itself) whereas "ctx" is actually more precise - special memory region named `ctx`.

dthaler replied:
Updated ctx comment.  I agree renames would be helpful in general.

elazarg commented on src/main/check.cpp:
```diff
@@ -79,6 +79,7 @@ int main(int argc, char** argv) {
         return 0;
     }
 
+    // Read a set of raw program sections from an ELF file.
```
Better one line below.

dthaler replied:
Moved

elazarg commented on src/asm_syntax.hpp:
```diff
@@ -14,17 +14,22 @@ using label_t = std::string;
 }
 using crab::label_t;
 
+// Abstract Syntax Machine (ASM) syntax.
```
I used `asm` as shorthand for "assembly" :) (as opposed to crab-related parts). If it's not obvious, then it's worth renaming too. Probably just `abstract_syntax` for this file.

The namespace is here mostly for doxygen to cluster the related classes together; that's why `using` follows immediately.

dthaler replied:
Updated comment

elazarg commented on src/crab/array_domain.cpp:
```diff
@@ -1,4 +1,4 @@
-
+// This file is eBPF-specific, not derived from CRAB.
```
This domain is implemented in Crab (there only in a header file: [domains/array_expansion.hpp](https://github.com/seahorn/crab/blob/master/include/crab/domains/array_expansion.hpp)) though it was designed specifically for eBPF-style register spilling. The implementation looks quite different due to extensive refactoring, but the logic is essentially the same. I'm not sure about licensing, but I did not implement the logic from scratch.

dthaler replied:
Got it, removed comment

elazarg commented:
@dthaler Looks good to me, Let me know when to rebase and merge. (Currently there are conflicts)

dthaler commented:
> @dthaler Looks good to me, Let me know when to rebase and merge. (Currently there are conflicts)

Rebased to resolve conflicts, and squashed comments that you previously reviewed.  Latest commit just has the two updates based on your latest code review comments.
